hazelcast/test/src/compact_test.cpp

```diff
@@ -0,0 +1,244 @@
+#include <hazelcast/client/exception/protocol_exceptions.h>
+#include <hazelcast/client/hazelcast_client.h>
+#include <hazelcast/client/hazelcast_json_value.h>
+#include <hazelcast/client/imap.h>
+#include <hazelcast/client/impl/Partition.h>
+#include <hazelcast/client/initial_membership_event.h>
+#include <hazelcast/client/internal/nearcache/impl/NearCacheRecordStore.h>
+#include <hazelcast/client/internal/nearcache/impl/store/NearCacheDataRecordStore.h>
+#include <hazelcast/client/internal/nearcache/impl/store/NearCacheObjectRecordStore.h>
+#include <hazelcast/client/internal/socket/SSLSocket.h>
+#include <hazelcast/client/lifecycle_listener.h>
+#include <hazelcast/client/pipelining.h>
+#include <hazelcast/client/serialization_config.h>
+#include <hazelcast/client/serialization/pimpl/data_input.h>
+#include <hazelcast/client/serialization/serialization.h>
+#include <hazelcast/util/AddressHelper.h>
+#include <hazelcast/util/concurrent/locks/LockSupport.h>
+#include <hazelcast/util/MurmurHash3.h>
+#include <hazelcast/util/Util.h>
+
+#include "ClientTest.h"
+#include "HazelcastServer.h"
+#include "HazelcastServerFactory.h"
+#include "TestHelperFunctions.h"
+#include "serialization/Serializables.h"
+#include "remote_controller_client.h"
+
+#if defined(WIN32) || defined(_WIN32) || defined(WIN64) || defined(_WIN64)
+#pragma warning(push)
+#pragma warning(disable : 4996) // for unsafe getenv
+#endif
+using namespace hazelcast::client::serialization::pimpl;
+namespace hazelcast {
+namespace client {
+namespace compact {
+namespace test {
+
+struct main_dto
+{
+    int i;
+    std::string str;
+};
+
+bool
+operator==(const main_dto& lhs, const main_dto& rhs)
+{
+    if (lhs.i != rhs.i) {
+        return false;
+    }
+    return lhs.str == rhs.str;
+}
+
+main_dto
+create_main_dto()
+{
+    return main_dto{ 30, "John" };
+}
+
+std::ostream&
+operator<<(std::ostream& out, const main_dto& main_dto)
+{
+    out << "i " << main_dto.i << ", str " << main_dto.str;
+    return out;
+}
+
+/**
+ * This class is to simulate versioning.
+ * We will provide this struct with serializer returning type name of the
+ * original main dto. This way we can use the serialized data of this class
+ * to test to_object of the original main_dto.
+ */
+struct empty_main_dto
+{};
+
+} // namespace test
+} // namespace compact
+
+namespace serialization {
+template<>
+struct hz_serializer<compact::test::main_dto> : public compact_serializer
+{
+    static void write(const compact::test::main_dto& object,
+                      compact_writer& writer)
+    {
+        writer.write_int32("i", object.i);
+        writer.write_string("name", object.str);
+    }
+
+    static compact::test::main_dto read(compact_reader& reader)
+    {
+        auto i = reader.read_int32("i", 1);
+        auto str =
+          reader.read_string("name", boost::make_optional<std::string>("NA"));
+        return compact::test::main_dto{ i, str.get_value_or("default") };
```

Comment:
I guess we can use on of the followings because we know there will be a value. If you want to keep using `get_value_or` I would change "default" to "NA" to avoid confusion of the default value.
```suggestion
        return compact::test::main_dto{ i, str.get() };
```

```suggestion
        return compact::test::main_dto{ i, *str };
```

Reply:
Fixed here 33186c83e8fcb88310e61706f80bbf9c8ea32528